content/blog/second-adventure-in-deno-land.md

```diff
@@ -0,0 +1,351 @@
+
+```html
+<script type="module">
+  // Imports the generated client
+  import * as client from "./client.js"
+
+  async function fetchTwitter(event) {
+    // Uses the methods on it
+    const result = await client.popular(event.target.value)
+
+    /*
+      Omitted for brevity
+    */
+  }
+</script>
+```
+
+[This is the file on GitHub](https://github.com/asantos00/deno-twitter-popular/blob/master/public/index.html)
+
+Which uses the client that was initially written in _deno_, and is now a js file.
+
+Now, this `public` folder can be served by any webserver. Since we're talking about _deno_, we can take advantage of it and serve the files with standard library's file server.
+
+```bash
+# inside the public folder
+$ deno run --allow-net --allow-read https://deno.land/std/http/file_server.ts
+```
+
+We can now visit `http://0.0.0.0:4507/` and use our very archaic frontend to query for popular tweets. With this, our code can use the client that was **originally written in deno** on the frontend, to interact with the API.
+
+![deno_tweets](./second-adventure-in-deno-land/deno_tweets.png)
+
+## Testing
+
+Following _deno_'s goal of shipping the essentials in a single binary, testing is obviously included. The documentation is, again, quite good on this. With the help of the `doc` command, it gets very straight-forward to write tests.
+
+```bash
+$ deno doc https://deno.land/std/testing/asserts.ts
+```
+
+This prints the documentation for the assert module from _deno_ standard library
+
+```ts
+function assert(expr: unknown, msg): <UNIMPLEMENTED>
+  Make an assertion, if not `true`, then throw.
+
+function assertEquals(actual: unknown, expected: unknown, msg?: string): void
+  Make an assertion that `actual` and `expected` are equal, deeply. If not deeply equal, then throw.
+
+function assertMatch(actual: string, expected: RegExp, msg?: string): void
+  Make an assertion that `actual` match RegExp `expected`. If not then thrown
+
+/* Cut for brevity */
+
+function fail(msg?: string): void
+  Forcefully throws a failed assertion
+
+```
+
+Let's write a simple test to our "isomorphic" client we created a few lines above.
+
+`Deno.test` will be used to declare the test body, `assertEquals` will be called to make the assertion. In order to mock the server responses, we'll spin up a server using standard-library _http_server_.
+
+We expect that, the `popular` method from the client calls `/popular/:twitterHandle`. Let's create a test for that.
+
+```ts
+import * as ApiClient from "./index.ts"
+import { assertEquals } from "../deps.ts"
+import { runServer } from "../util.ts"
+
+Deno.test("calls the correct url", async () => {
+  runServer(async req => {
+    assertEquals(req.url, "/popular/ampsantos0")
+    await req.respond({ body: JSON.stringify({ statuses: [] }) })
+  })
+
+  await ApiClient.popular("ampsantos0")
+})
+```
+
+Our `runServer` util is a very simple function that just spawns a web server that closes after the first connection.
```

Comment:
Might be cool on a future adventure to also explore stuff like [mocks](https://deno.land/x/mock/), etc.

Reply:
It is, definitely! I'll probably do a third adventure with benchmarking and using a couple of libraries to very standard use cases. I was thinking like `oak` (as you talked about last time), `mock` and a couple more things